From f4c9df7512871f437c2d123c7916f3d85bc90600 Mon Sep 17 00:00:00 2001
From: Olivier Dugas <dugaso@sonatest.com>
Date: Tue, 18 Jul 2017 13:32:36 -0400
Subject: [PATCH 201/532] iio: adc: xilinx-ams: reset [ps|pl]-sysmon before
 channel config

commit bdd66cbb5ab7829338bbbbb930b70cb386d9825b from
https://github.com/Xilinx/linux-xlnx.git

The documentation for PS and PL System Monitor Programming Model
specifies that in order to change which channel to monitor, the SYSMON
must be put in a soft reset. This is done by modifying its
sequence_mode to default.

This fixes in_voltage15_vccbram_raw always reading 0.

Signed-off-by: Jean-Francois Dagenais <jeff.dagenais@gmail.com>
Signed-off-by: Olivier Dugas <dugaso@sonatest.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/iio/adc/xilinx-ams.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/iio/adc/xilinx-ams.c b/drivers/iio/adc/xilinx-ams.c
index 30a928e..d083264 100644
--- a/drivers/iio/adc/xilinx-ams.c
+++ b/drivers/iio/adc/xilinx-ams.c
@@ -193,8 +193,13 @@ static void iio_ams_init_device(struct ams *ams)
 		scan_mask |= BIT(indio_dev->channels[i].scan_index);
 
 	if (ams->ps_base) {
+		/* put sysmon in a soft reset to change the sequence */
+		ams_ps_update_reg(ams, AMS_REG_CONFIG1, AMS_CONF1_SEQ_MASK,
+				  AMS_CONF1_SEQ_DEFAULT);
+
+		/* configure basic channels */
 		ams_ps_write_reg(ams, AMS_REG_SEQ_CH0,
-				scan_mask & AMS_REG_SEQ0_MASK);
+				 scan_mask & AMS_REG_SEQ0_MASK);
 		ams_ps_write_reg(ams, AMS_REG_SEQ_CH2, AMS_REG_SEQ2_MASK &
 				(scan_mask >> AMS_REG_SEQ2_MASK_SHIFT));
 
@@ -204,6 +209,11 @@ static void iio_ams_init_device(struct ams *ams)
 	}
 
 	if (ams->pl_base) {
+		/* put sysmon in a soft reset to change the sequence */
+		ams->pl_bus->update(ams, AMS_REG_CONFIG1, AMS_CONF1_SEQ_MASK,
+				    AMS_CONF1_SEQ_DEFAULT);
+
+		/* configure basic channels */
 		scan_mask = (scan_mask >> PS_SEQ_MAX);
 		ams->pl_bus->write(ams, AMS_REG_SEQ_CH0,
 				scan_mask & AMS_REG_SEQ0_MASK);
-- 
2.7.4

